File: packages/python/elizaos/advanced_capabilities/form/storage.py
"""Component-based persistence for form data.

Form data is stored using elizaOS's Component system:
- Sessions stored as ``form_session:{roomId}``
- Submissions stored as ``form_submission:{formId}:{submissionId}``
- Autofill stored as ``form_autofill:{formId}``
"""

from __future__ import annotations

import json
from typing import TYPE_CHECKING, Any

from .types import (
    FORM_AUTOFILL_COMPONENT,
    FORM_SESSION_COMPONENT,
    FORM_SUBMISSION_COMPONENT,
    FormAutofillData,
    FormSession,
    FormSubmission,
)

if TYPE_CHECKING:
    from elizaos.types import IAgentRuntime


def _session_component_type(room_id: str) -> str:
    return f"{FORM_SESSION_COMPONENT}:{room_id}"


def _submission_component_type(form_id: str, submission_id: str) -> str:
    return f"{FORM_SUBMISSION_COMPONENT}:{form_id}:{submission_id}"


def _autofill_component_type(form_id: str) -> str:
    return f"{FORM_AUTOFILL_COMPONENT}:{form_id}"


def _session_to_dict(session: FormSession) -> dict[str, Any]:
    """Serialize a FormSession to a JSON-compatible dict."""
    import dataclasses

    return json.loads(json.dumps(dataclasses.asdict(session), default=str))


def _dict_to_session(data: dict[str, Any]) -> FormSession:
    """Deserialize a dict to a FormSession (best-effort)."""
    from .types import FieldHistoryEntry, FieldState, SessionEffort

    fields: dict[str, FieldState] = {}
    for k, v in data.get("fields", {}).items():
        if isinstance(v, dict):
            fields[k] = FieldState(
                **{fk: fv for fk, fv in v.items() if fk in FieldState.__dataclass_fields__}
            )
        else:
            fields[k] = FieldState()

    history: list[FieldHistoryEntry] = []
    for entry in data.get("history", []):
        if isinstance(entry, dict):
            history.append(
                FieldHistoryEntry(
                    field=entry.get("field", ""),
                    old_value=entry.get("old_value"),
                    new_value=entry.get("new_value"),
                    timestamp=entry.get("timestamp", 0),
                )
            )

    effort_data = data.get("effort", {})
    effort = SessionEffort(
        interaction_count=effort_data.get("interaction_count", 0),
        time_spent_ms=effort_data.get("time_spent_ms", 0),
        first_interaction_at=effort_data.get("first_interaction_at", 0),
        last_interaction_at=effort_data.get("last_interaction_at", 0),
    )

    return FormSession(
        id=data.get("id", ""),
        form_id=data.get("form_id", ""),
        entity_id=data.get("entity_id", ""),
        room_id=data.get("room_id", ""),
        status=data.get("status", "active"),
        form_version=data.get("form_version"),
        fields=fields,
        history=history,
        parent_session_id=data.get("parent_session_id"),
        context=data.get("context"),
        locale=data.get("locale"),
        last_asked_field=data.get("last_asked_field"),
        last_message_id=data.get("last_message_id"),
        cancel_confirmation_asked=data.get("cancel_confirmation_asked", False),
        effort=effort,
        expires_at=data.get("expires_at", 0),
        expiration_warned=data.get("expiration_warned", False),
        nudge_count=data.get("nudge_count", 0),
        last_nudge_at=data.get("last_nudge_at"),
        created_at=data.get("created_at", 0),
        updated_at=data.get("updated_at", 0),
        submitted_at=data.get("submitted_at"),
        meta=data.get("meta"),
    )


# ---------------------------------------------------------------------------
# Session operations
# ---------------------------------------------------------------------------


async def get_active_session(
    runtime: IAgentRuntime,
    entity_id: str,
    room_id: str,
) -> FormSession | None:
    """Get active form session for entity in a specific room."""
    component_type = _session_component_type(room_id)
    components = await runtime.get_components(entity_id, component_type)
    for comp in components:
        data = comp.get("data", {})
        if isinstance(data, dict) and data.get("status") in ("active", "ready"):
            return _dict_to_session(data)
    return None


async def get_all_active_sessions(
    runtime: IAgentRuntime,
    entity_id: str,
) -> list[FormSession]:
    """Get all active sessions for an entity across all rooms."""
    components = await runtime.get_components(entity_id, FORM_SESSION_COMPONENT)
    sessions: list[FormSession] = []
    for comp in components:
        data = comp.get("data", {})
        if isinstance(data, dict) and data.get("status") in ("active", "ready"):
            sessions.append(_dict_to_session(data))
    return sessions


async def get_stashed_sessions(
    runtime: IAgentRuntime,
    entity_id: str,
) -> list[FormSession]:
    """Get stashed sessions for an entity."""
    components = await runtime.get_components(entity_id, FORM_SESSION_COMPONENT)
    sessions: list[FormSession] = []
    for comp in components:
        data = comp.get("data", {})
        if isinstance(data, dict) and data.get("status") == "stashed":
            sessions.append(_dict_to_session(data))
    return sessions


async def get_session_by_id(
    runtime: IAgentRuntime,
    entity_id: str,
    session_id: str,
) -> FormSession | None:
    """Get a session by its ID."""
    components = await runtime.get_components(entity_id, FORM_SESSION_COMPONENT)
    for comp in components:
        data = comp.get("data", {})
        if isinstance(data, dict) and data.get("id") == session_id:
            return _dict_to_session(data)
    return None


async def save_session(
    runtime: IAgentRuntime,
    session: FormSession,
) -> None:
    """Save a form session (upsert)."""
    component_type = _session_component_type(session.room_id)
    data = _session_to_dict(session)
    await runtime.set_component(
        session.entity_id,
        component_type,
        data,
    )


async def delete_session(
    runtime: IAgentRuntime,
    session: FormSession,
) -> None:
    """Delete a session."""
    component_type = _session_component_type(session.room_id)
    await runtime.delete_component(session.entity_id, component_type)


# ---------------------------------------------------------------------------
# Submission operations
# ---------------------------------------------------------------------------


async def save_submission(
    runtime: IAgentRuntime,
    submission: FormSubmission,
) -> None:
    """Save a form submission (immutable record)."""
    import dataclasses

    component_type = _submission_component_type(submission.form_id, submission.id)
    data = json.loads(json.dumps(dataclasses.asdict(submission), default=str))
    await runtime.set_component(
        submission.entity_id,
        component_type,
        data,
    )


async def get_submissions(
    runtime: IAgentRuntime,
    entity_id: str,
    form_id: str | None = None,
) -> list[FormSubmission]:
    """Get submissions for an entity, optionally filtered by form ID."""
    prefix = FORM_SUBMISSION_COMPONENT
    if form_id:
        prefix = f"{FORM_SUBMISSION_COMPONENT}:{form_id}"
    components = await runtime.get_components(entity_id, prefix)
    results: list[FormSubmission] = []
    for comp in components:
        data = comp.get("data", {})
        if isinstance(data, dict):
            results.append(
                FormSubmission(
                    id=data.get("id", ""),
                    form_id=data.get("form_id", ""),
                    session_id=data.get("session_id", ""),
                    entity_id=data.get("entity_id", ""),
                    values=data.get("values", {}),
                    form_version=data.get("form_version"),
                    mapped_values=data.get("mapped_values"),
                    files=data.get("files"),
                    submitted_at=data.get("submitted_at", 0),
                    meta=data.get("meta"),
                )
            )
    return sorted(results, key=lambda s: s.submitted_at, reverse=True)


async def get_submission_by_id(
    runtime: IAgentRuntime,
    entity_id: str,
    submission_id: str,
) -> FormSubmission | None:
    """Get a specific submission by ID."""
    subs = await get_submissions(runtime, entity_id)
    for s in subs:
        if s.id == submission_id:
            return s
    return None


# ---------------------------------------------------------------------------
# Autofill operations
# ---------------------------------------------------------------------------


async def get_autofill_data(
    runtime: IAgentRuntime,
    entity_id: str,
    form_id: str,
) -> FormAutofillData | None:
    """Get autofill data for a user's form."""
    component_type = _autofill_component_type(form_id)
    components = await runtime.get_components(entity_id, component_type)
    for comp in components:
        data = comp.get("data", {})
        if isinstance(data, dict):
            return FormAutofillData(
                form_id=data.get("form_id", form_id),
                values=data.get("values", {}),
                updated_at=data.get("updated_at", 0),
            )
    return None


async def save_autofill_data(
    runtime: IAgentRuntime,
    entity_id: str,
    form_id: str,
    values: dict[str, Any],
) -> None:
    """Save autofill data for a user's form."""
    import time

    component_type = _autofill_component_type(form_id)
    data = {
        "form_id": form_id,
        "values": values,
        "updated_at": int(time.time() * 1000),
    }
    await runtime.set_component(entity_id, component_type, data)


# ---------------------------------------------------------------------------
# Placeholder: stale/expiring session queries
# ---------------------------------------------------------------------------


async def get_stale_sessions(
    runtime: IAgentRuntime,
    after_inactive_ms: int,
) -> list[FormSession]:
    """Placeholder -- requires database-level query in production."""
    _ = runtime, after_inactive_ms
    return []


async def get_expiring_sessions(
    runtime: IAgentRuntime,
    within_ms: int,
) -> list[FormSession]:
    """Placeholder -- requires database-level query in production."""
    _ = runtime, within_ms
    return []

File: apps/app-form/src/storage.ts
/**
 * @module storage
 * @description Component-based persistence for form data
 *
 * ## Design Rationale
 *
 * Form data is stored using elizaOS's Component system because:
 *
 * 1. **Entity-Scoped**: Components belong to entities (users).
 *    This naturally scopes form data per-user.
 *
 * 2. **Typed Storage**: Component type field allows different kinds
 *    of form data (sessions, submissions, autofill).
 *
 * 3. **No Custom Schema**: Uses existing elizaOS infrastructure,
 *    no need to create database tables.
 *
 * 4. **Room Scoping**: Component type includes roomId for session
 *    isolation across rooms.
 *
 * ## Storage Strategy
 *
 * ### Sessions
 * - Stored as components with type: `form_session:{roomId}`
 * - One active session per user per room
 * - Scoping ensures different rooms have different contexts
 *
 * ### Submissions
 * - Stored as components with type: `form_submission:{formId}:{submissionId}`
 * - Immutable records of completed forms
 * - Multiple submissions per user (if form allows)
 *
 * ### Autofill
 * - Stored as components with type: `form_autofill:{formId}`
 * - One autofill record per user per form
 * - Updated on each submission
 *
 * ## Limitations
 *
 * The current implementation has limitations:
 *
 * 1. **No Cross-Entity Queries**: Can't efficiently find all stale
 *    sessions across all users. This affects nudge system.
 *
 * 2. **No Indexes**: Component queries are sequential scans.
 *    For high-volume usage, consider database-level optimizations.
 *
 * These are acceptable for v1 but noted for future improvement.
 */

import type { Component, IAgentRuntime, JsonValue, UUID } from "@elizaos/core";
import { v4 as uuidv4 } from "uuid";
import type { FormAutofillData, FormSession, FormSubmission } from "./types";
import {
  FORM_AUTOFILL_COMPONENT,
  FORM_SESSION_COMPONENT,
  FORM_SUBMISSION_COMPONENT,
} from "./types";

const isRecord = (value: JsonValue | object): value is Record<string, JsonValue> =>
  typeof value === "object" && value !== null && !Array.isArray(value);

const resolveComponentContext = async (
  runtime: IAgentRuntime,
  roomId?: UUID
): Promise<{ roomId: UUID; worldId: UUID }> => {
  if (roomId) {
    const room = await runtime.getRoom(roomId);
    return { roomId, worldId: room?.worldId ?? runtime.agentId };
  }
  return { roomId: runtime.agentId, worldId: runtime.agentId };
};

const isFormSession = (data: JsonValue | object): data is FormSession => {
  if (!isRecord(data)) return false;
  return (
    typeof data.id === "string" &&
    typeof data.formId === "string" &&
    typeof data.entityId === "string" &&
    typeof data.roomId === "string"
  );
};

const isFormSubmission = (data: JsonValue | object): data is FormSubmission => {
  if (!isRecord(data)) return false;
  return (
    typeof data.id === "string" &&
    typeof data.formId === "string" &&
    typeof data.sessionId === "string" &&
    typeof data.entityId === "string"
  );
};

const isFormAutofillData = (data: JsonValue | object): data is FormAutofillData => {
  if (!isRecord(data)) return false;
  return (
    typeof data.formId === "string" &&
    typeof data.updatedAt === "number" &&
    typeof data.values === "object"
  );
};

// ============================================================================
// SESSION STORAGE
// ============================================================================

/**
 * Get active form session for entity in a specific room.
 *
 * WHY room-scoped:
 * - User might chat in multiple rooms simultaneously
 * - Each room conversation should have its own form context
 * - Discord DM form shouldn't interfere with Telegram form
 *
 * WHY active/ready filter:
 * - Stashed, submitted, cancelled, expired sessions are not "active"
 * - User would need to restore stashed sessions
 *
 * @param runtime - Agent runtime for database access
 * @param entityId - User's entity ID
 * @param roomId - The room to check for active session
 * @returns Active session or null if none
 */
export async function getActiveSession(
  runtime: IAgentRuntime,
  entityId: UUID,
  roomId: UUID
): Promise<FormSession | null> {
  // Component type includes roomId for room-level scoping
  const component = await runtime.getComponent(entityId, `${FORM_SESSION_COMPONENT}:${roomId}`);

  if (!component?.data || !isFormSession(component.data)) return null;

  const session = component.data;

  // Only return if active (not stashed, submitted, cancelled, or expired)
  // WHY: Other statuses require explicit action to restore/continue
  if (session.status === "active" || session.status === "ready") {
    return session;
  }

  return null;
}

/**
 * Get all active sessions for an entity (across all rooms).
 *
 * WHY this exists:
 * - For "you have forms in progress" notifications
 * - For session management UI
 * - Not commonly used in normal flow
 *
 * @param runtime - Agent runtime for database access
 * @param entityId - User's entity ID
 * @returns Array of active sessions (may be empty)
 */
export async function getAllActiveSessions(
  runtime: IAgentRuntime,
  entityId: UUID
): Promise<FormSession[]> {
  const components = await runtime.getComponents(entityId);

  const sessions: FormSession[] = [];
  for (const component of components) {
    // Check if this is a form session component
    if (component.type.startsWith(`${FORM_SESSION_COMPONENT}:`)) {
      if (component.data && isFormSession(component.data)) {
        const session = component.data;
        if (session.status === "active" || session.status === "ready") {
          sessions.push(session);
        }
      }
    }
  }

  return sessions;
}

/**
 * Get stashed sessions for an entity.
 *
 * WHY stashed is separate from active:
 * - Stashed sessions are "saved for later"
 * - User must explicitly restore them
 * - Different UX from active sessions
 *
 * @param runtime - Agent runtime for database access
 * @param entityId - User's entity ID
 * @returns Array of stashed sessions (may be empty)
 */
export async function getStashedSessions(
  runtime: IAgentRuntime,
  entityId: UUID
): Promise<FormSession[]> {
  const components = await runtime.getComponents(entityId);

  const sessions: FormSession[] = [];
  for (const component of components) {
    if (component.type.startsWith(`${FORM_SESSION_COMPONENT}:`)) {
      if (component.data && isFormSession(component.data)) {
        const session = component.data;
        if (session.status === "stashed") {
          sessions.push(session);
        }
      }
    }
  }

  return sessions;
}

/**
 * Get a session by ID.
 *
 * WHY by ID:
 * - Needed for operations on specific session
 * - Session ID is stable across room changes
 * - Used by stash/restore when session roomId changes
 *
 * @param runtime - Agent runtime for database access
 * @param entityId - User's entity ID
 * @param sessionId - The session ID to find
 * @returns The session or null if not found
 */
export async function getSessionById(
  runtime: IAgentRuntime,
  entityId: UUID,
  sessionId: string
): Promise<FormSession | null> {
  const components = await runtime.getComponents(entityId);

  for (const component of components) {
    if (component.type.startsWith(`${FORM_SESSION_COMPONENT}:`)) {
      if (component.data && isFormSession(component.data)) {
        const session = component.data;
        if (session.id === sessionId) {
          return session;
        }
      }
    }
  }

  return null;
}

/**
 * Save a form session.
 *
 * Creates new component if none exists, updates otherwise.
 *
 * WHY upsert pattern:
 * - Session is created once, updated many times
 * - Single function handles both cases
 * - Avoids race conditions
 *
 * @param runtime - Agent runtime for database access
 * @param session - Session to save
 */
export async function saveSession(runtime: IAgentRuntime, session: FormSession): Promise<void> {
  const componentType = `${FORM_SESSION_COMPONENT}:${session.roomId}`;
  const existing = await runtime.getComponent(session.entityId, componentType);
  const context = await resolveComponentContext(runtime, session.roomId);
  const resolvedWorldId = existing?.worldId ?? context.worldId;

  const component: Component = {
    id: existing?.id || (uuidv4() as UUID),
    entityId: session.entityId,
    agentId: runtime.agentId,
    roomId: session.roomId,
    // WHY preserve worldId: Avoids breaking existing component relationships
    worldId: resolvedWorldId,
    sourceEntityId: runtime.agentId,
    type: componentType,
    createdAt: existing?.createdAt || Date.now(),
    // Store session as component data
    data: JSON.parse(JSON.stringify(session)) as Record<string, JsonValue>,
  };

  if (existing) {
    await runtime.updateComponent(component);
  } else {
    await runtime.createComponent(component);
  }
}

/**
 * Delete a session.
 *
 * WHY delete:
 * - Cleanup after submission/cancellation/expiry
 * - Frees up storage
 * - Note: Usually we just change status instead of deleting
 *
 * @param runtime - Agent runtime for database access
 * @param session - Session to delete
 */
export async function deleteSession(runtime: IAgentRuntime, session: FormSession): Promise<void> {
  const componentType = `${FORM_SESSION_COMPONENT}:${session.roomId}`;
  const existing = await runtime.getComponent(session.entityId, componentType);

  if (existing) {
    await runtime.deleteComponent(existing.id);
  }
}

// ============================================================================
// SUBMISSION STORAGE
// ============================================================================

/**
 * Save a form submission.
 *
 * Submissions are immutable records. Always creates new component.
 *
 * WHY new component per submission:
 * - Submissions are immutable
 * - Multiple submissions allowed (if form permits)
 * - Historical record keeping
 *
 * @param runtime - Agent runtime for database access
 * @param submission - Submission to save
 */
export async function saveSubmission(
  runtime: IAgentRuntime,
  submission: FormSubmission
): Promise<void> {
  // Use a unique component type per submission
  // WHY: Allows multiple submissions per form
  const componentType = `${FORM_SUBMISSION_COMPONENT}:${submission.formId}:${submission.id}`;
  const context = await resolveComponentContext(runtime);

  const component: Component = {
    id: uuidv4() as UUID,
    entityId: submission.entityId,
    agentId: runtime.agentId,
    roomId: context.roomId,
    worldId: context.worldId,
    sourceEntityId: runtime.agentId,
    type: componentType,
    createdAt: submission.submittedAt,
    data: JSON.parse(JSON.stringify(submission)) as Record<string, JsonValue>,
  };

  await runtime.createComponent(component);
}

/**
 * Get submissions for an entity, optionally filtered by form ID.
 *
 * WHY optional formId:
 * - List all submissions: no formId
 * - List submissions for specific form: with formId
 *
 * @param runtime - Agent runtime for database access
 * @param entityId - User's entity ID
 * @param formId - Optional form ID filter
 * @returns Array of submissions, newest first
 */
export async function getSubmissions(
  runtime: IAgentRuntime,
  entityId: UUID,
  formId?: string
): Promise<FormSubmission[]> {
  const components = await runtime.getComponents(entityId);

  const submissions: FormSubmission[] = [];
  const prefix = formId
    ? `${FORM_SUBMISSION_COMPONENT}:${formId}:`
    : `${FORM_SUBMISSION_COMPONENT}:`;

  for (const component of components) {
    if (component.type.startsWith(prefix)) {
      if (component.data && isFormSubmission(component.data)) {
        submissions.push(component.data);
      }
    }
  }

  // Sort by submission time, newest first
  // WHY: Most recent submissions are usually most relevant
  submissions.sort((a, b) => b.submittedAt - a.submittedAt);

  return submissions;
}

/**
 * Get a specific submission by ID.
 *
 * @param runtime - Agent runtime for database access
 * @param entityId - User's entity ID
 * @param submissionId - The submission ID to find
 * @returns The submission or null if not found
 */
export async function getSubmissionById(
  runtime: IAgentRuntime,
  entityId: UUID,
  submissionId: string
): Promise<FormSubmission | null> {
  const components = await runtime.getComponents(entityId);

  for (const component of components) {
    if (component.type.startsWith(`${FORM_SUBMISSION_COMPONENT}:`)) {
      if (component.data && isFormSubmission(component.data)) {
        const submission = component.data;
        if (submission.id === submissionId) {
          return submission;
        }
      }
    }
  }

  return null;
}

// ============================================================================
// AUTOFILL STORAGE
// ============================================================================

/**
 * Get autofill data for a user's form.
 *
 * WHY autofill:
 * - Users filling repeat forms want saved values
 * - Reduces friction for common fields (name, email, address)
 *
 * @param runtime - Agent runtime for database access
 * @param entityId - User's entity ID
 * @param formId - Form definition ID
 * @returns Autofill data or null if none saved
 */
export async function getAutofillData(
  runtime: IAgentRuntime,
  entityId: UUID,
  formId: string
): Promise<FormAutofillData | null> {
  const componentType = `${FORM_AUTOFILL_COMPONENT}:${formId}`;
  const component = await runtime.getComponent(entityId, componentType);

  if (!component?.data || !isFormAutofillData(component.data)) return null;

  return component.data;
}

/**
 * Save autofill data for a user's form.
 *
 * Overwrites existing autofill data for the form.
 *
 * WHY overwrite:
 * - Most recent submission has most current data
 * - User's email might have changed
 * - Only one autofill record per form needed
 *
 * @param runtime - Agent runtime for database access
 * @param entityId - User's entity ID
 * @param formId - Form definition ID
 * @param values - Field values to save for autofill
 */
export async function saveAutofillData(
  runtime: IAgentRuntime,
  entityId: UUID,
  formId: string,
  values: Record<string, JsonValue>
): Promise<void> {
  const componentType = `${FORM_AUTOFILL_COMPONENT}:${formId}`;
  const existing = await runtime.getComponent(entityId, componentType);
  const context = await resolveComponentContext(runtime);
  const resolvedWorldId = existing?.worldId ?? context.worldId;

  const data: FormAutofillData = {
    formId,
    values,
    updatedAt: Date.now(),
  };

  const component: Component = {
    id: existing?.id || (uuidv4() as UUID),
    entityId,
    agentId: runtime.agentId,
    roomId: context.roomId,
    worldId: resolvedWorldId,
    sourceEntityId: runtime.agentId,
    type: componentType,
    createdAt: existing?.createdAt || Date.now(),
    data: JSON.parse(JSON.stringify(data)) as Record<string, JsonValue>,
  };

  if (existing) {
    await runtime.updateComponent(component);
  } else {
    await runtime.createComponent(component);
  }
}

// ============================================================================
// QUERY HELPERS
// ============================================================================

/**
 * Get all stale sessions (for nudge system).
 *
 * LIMITATION: This requires iterating over all entities, which is not
 * efficient with current elizaOS component system. In production,
 * this would need a database-level query.
 *
 * WHY this is here:
 * - Documents the need for this functionality
 * - Placeholder for future implementation
 * - Nudge system can call this when entity context is available
 *
 * @param runtime - Agent runtime for database access
 * @param afterInactiveMs - Inactivity threshold in milliseconds
 * @returns Array of stale sessions (currently returns empty array)
 */
export async function getStaleSessions(
  runtime: IAgentRuntime,
  _afterInactiveMs: number
): Promise<FormSession[]> {
  // Proper querying across all entities would require either a database index
  // on component data, a separate tracking table, or a periodic full scan.

  runtime.logger.warn("getStaleSessions requires entity iteration - not implemented");
  return [];
}

/**
 * Get sessions expiring within a time window.
 *
 * Same limitation as getStaleSessions.
 *
 * @param runtime - Agent runtime for database access
 * @param withinMs - Time window in milliseconds
 * @returns Array of expiring sessions (currently returns empty array)
 */
export async function getExpiringSessions(
  runtime: IAgentRuntime,
  _withinMs: number
): Promise<FormSession[]> {
  runtime.logger.warn("getExpiringSessions requires entity iteration - not implemented");
  return [];
}
